### PR TITLE
Add a proper dependency factory API for dependencies with more than one discovery method

### PR DESCRIPTION
**Pull request by dcbaker**
This adds a more compact, explict method for dependencies with more than one way to be found, especially when one or more of those methods are standard, built in methods like cmake and pkg-config. Instead of having a special `_factory` method in the dependency, a new generic `DependencyFactory` has been added that can handle many cases, or a specific factory function can be written for more specialized cases. I've included a number of instances of both.

I've converted most instances that make sense, although there are a few more that could use conversion I haven't gotten to: Qt, scalapack, and mpi (maybe a couple of others I forgot about). I haven't had a chance to test this on mac or windows yet, so it may still have some problems.

@scivision, I've touched several of the dependencies you've added without any way to test them (I can' seem to get ifort set up correctly with MPI for some reason)

### CODE REVIEW

**Comment by ignatenkobrain** [comment]
Flake8 detected 7 issues on the commit b139e1245ba289e3055db256574217c9ddfba2ce.
Visit the [analysis result](https://sider.review/gh/repos/19784232/pulls/6443) to review all issues.

<sub>posted by Sider</sub>

File: mesonbuild/dependencies/ui.py [hunk 1]
```diff
@@ -30,12 +32,12 @@
 from .base import DependencyException, DependencyMethods
 from .base import ExternalDependency, ExternalProgram, NonExistingExternalProgram
 from .base import ExtraFrameworkDependency, PkgConfigDependency
-from .base import ConfigToolDependency
+from .base import ConfigToolDependency, process_method_kw, DependencyFactory
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/ui.py]: __[Flake8]__ '.base.process_method_kw' imported but unused ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6443#issue-5055575))

<details>

| Rule |
| - |
| F401 |

</details>
<sub>posted by Sider</sub>

File: mesonbuild/dependencies/dev.py [hunk 2]
```diff
@@ -391,17 +391,19 @@ def log_details(self):
...
-        super().__init__(name='LLVM', environment=env, language='cpp', kwargs=kwargs)
+        super().__init__(name, env, kwargs, language='cpp')
 
         # Extract extra include directories and definitions
         inc_dirs = self.traceparser.get_cmake_var('PACKAGE_INCLUDE_DIRS')
         defs = self.traceparser.get_cmake_var('PACKAGE_DEFINITIONS')
         temp = ['-I' + x for x in inc_dirs] + defs
         self.compile_args += [x for x in temp if x not in self.compile_args]
-        self._add_sub_dependency(ThreadDependency, env, kwargs)
+        if not self._add_sub_dependency(threads_factory(environment, self.for_machine, {})):
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/dev.py]: __[Flake8]__ undefined name 'environment' ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6443#issue-5055573))

<details>

| Rule |
| - |
| F821 |

</details>
<sub>posted by Sider</sub>

File: mesonbuild/dependencies/ui.py [hunk 3]
```diff
@@ -18,7 +18,9 @@
 import os
 import re
 import subprocess
+import typing as T
 from collections import OrderedDict
+import typing as T
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/ui.py]: __[Flake8]__ redefinition of unused 'T' from line 21 ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6443#issue-5055574))

<details>

| Rule |
| - |
| F811 |

</details>
<sub>posted by Sider</sub>

File: mesonbuild/dependencies/misc.py [hunk 4]
```diff
@@ -27,55 +28,38 @@
...
     DependencyException, DependencyMethods, ExternalDependency,
     ExtraFrameworkDependency, PkgConfigDependency,
-    CMakeDependency, ConfigToolDependency,
+    CMakeDependency, ConfigToolDependency, process_method_kw,
+    DependencyFactory,
 )
 
+if T.TYPE_CHECKING:
+    from ..environment import Environment, MachineChoice
+    from .base import DependencyType
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__ '.base.DependencyType' imported but unused ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6443#issue-5055572))

<details>

| Rule |
| - |
| F401 |

</details>
<sub>posted by Sider</sub>

**Comment by scivision** [comment]
For Intel MPI, "Intel Parallel Studio XE Cluster Edition" is required--the non-Cluster Edition don't have MPI. 

I opened the graphical installer now on Linux and Windows to copy down the options I have selected on each (either work fine for Meson testing). I put the slightly different names for Windows in curly brackets.

* Intel C++ Compiler 19.1 {for Intel 64}
* Intel Fortran Compiler 19.1 {Visual Fortran Compiler for Intel 64}
* Intel Math Kernel Library 2020 for C/C++ {for Intel 64}
  * Intel MKL core libraries for C/C++ {for Intel 64}
  * **Cluster support for C/C++**
* Intel Math Kernel Library 2020 for Fortran
  * Intel MKL core libraries for Fortran {for Intel 64}
  * **Cluster support for Fortran**
* Intel Threaded Building Blocks 2020 {TBB core libraries for Intel 64}
* Intel MPI Library 2019 Update 6

I emphasized the "cluster" options just to note those bring in the MPI numerical libraries one would test with Meson. The options are similar for previous years' Intel releases.

Intel Linux compilers also work fine from Windows Subsystem for Linux.

**Comment by scivision** [comment]
I think this is generally a good thing. I have a nearly complete PR to use CMake Components that will allow many more libraries to be used from Meson. Those libraries will use this PR's factory functionality to do that more cleanly.

**Comment by dcbaker** [comment]
I work for Intel, so I got the whole kitten-kaboodle :). I think there's something wrong witht he way I'm sourcing the scripts though, as there are some things in my PATH that are clearly supposed to be variables, but are not, things like `MPI_TMP_ROOT_PATH/mpi/...` (I made that up, but something like that).

Also, I can't replicate the macOS Qt failures, Qt works fine for me on my mac :/

File: mesonbuild/dependencies/base.py [hunk 5]
```diff
@@ -248,7 +248,7 @@ def get_for_machine_from_kwargs(self, kwargs):
         return MachineChoice.BUILD if kwargs.get('native', False) else MachineChoice.HOST
 
 class ExternalDependency(Dependency, HasNativeKwarg):
-    def __init__(self, type_name, environment, language, kwargs):
+    def __init__(self, type_name, environment, kwargs, language: T.Optional[str] = None):
```
Comment (jon-turney) [inline mesonbuild/dependencies/base.py]: Does the commentary for the commit containing this change want to say "Dependency constructor signatures" ?

File: mesonbuild/dependencies/base.py [hunk 6]
```diff
@@ -2137,6 +2140,79 @@ def log_tried(self):
...
 
 
+class DependencyFactory:
+
+    """Factory to get dependencies from multiple sources.
+
+    This class provides a constructor that takes a set of names and classes
+    for various kinds of dependencies. This simplifies getting a single
+    dependency from multiple sources.
+
```
Comment (jon-turney) [inline mesonbuild/dependencies/base.py]: (something like) "... calling the object returned by the constructor returns a list of constructors for Dependency-subclass objects to be tried in order" ?

File: mesonbuild/dependencies/base.py [hunk 7]
```diff
@@ -2137,6 +2140,79 @@ def log_tried(self):
...
+            return False
+        return True
+
+    def __call__(self, env: Environment, for_machine: MachineChoice,
+                 kwargs: T.Dict[str, T.Any]) -> T.List['DependencyType']:
+        """Return a list of Dependencies with the arguments already attached."""
+        methods = process_method_kw(self.methods, kwargs)
+        possible = [m for m in self.methods if self._process_method(m, env, for_machine)]
+
+        return [functools.partial(self.classes[e], env, kwargs) for e in possible if e in methods]
```
Comment (jon-turney) [inline mesonbuild/dependencies/base.py]: Writing this as a sequence of list comprehensions looks a bit odd. I think this is just testing if the method is selected (via the method: keyword, and possible), so could both could just be tested in the condition?)

File: mesonbuild/dependencies/misc.py [hunk 8]
```diff
@@ -222,7 +222,7 @@ def _factory(cls, environment, kwargs):
             _kargs = kwargs.copy()
             _kargs[paths] = ['/Library/Frameworks']
             candidates.append(functools.partial(
-                ExtraFrameworkDependency, 'Python', environment, _kargs)
+                ExtraFrameworkDependency, 'Python', environment, _kargs))
```
Comment (jon-turney) [inline mesonbuild/dependencies/misc.py]: Hmm... unrelated fix for a typo in previous commit?

File: mesonbuild/dependencies/base.py [hunk 9]
```diff
@@ -179,6 +179,22 @@ def get_partial_dependency(self, *, compile_args: bool = False,
         """
         raise RuntimeError('Unreachable code in partial_dependency called')
 
+    def _add_sub_dependency2(self, deplist: T.List['DependencyType']) -> bool:
+        """Add an internal depdency from a list of possible dependencies.
+
+        This method is inteneded to make it easier to add additional
```
Comment (jon-turney) [inline mesonbuild/dependencies/base.py]: "intended"

File: mesonbuild/dependencies/misc.py [hunk 10]
```diff
@@ -477,6 +457,41 @@ def get_methods():
         return [DependencyMethods.AUTO, DependencyMethods.PKGCONFIG]
 
 
+def shaderc_factory(self, env: 'Environment', for_machine: 'MachineChoice',
+             kwargs: T.Dict[str, T.Any]) -> T.List['DependencyType']:
+    """Custom DependencyFactory for ShaderC.
+
+    ShaderC's odd you get thhree different libraries from the same build
```
Comment (jon-turney) [inline mesonbuild/dependencies/misc.py]: "odd, in that you get three"

File: mesonbuild/dependencies/misc.py [hunk 11]
```diff
@@ -477,6 +457,41 @@ def get_methods():
...
+    thing are just easier too represent as a separate class than
```
Comment (jon-turney) [inline mesonbuild/dependencies/misc.py]: "this is just easier to represent"

File: mesonbuild/dependencies/base.py [hunk 12]
```diff
@@ -2137,6 +2140,79 @@ def log_tried(self):
...
+
+        self.methods = methods
+        self.classes = {
+            # Just attach the correct name right now, either the generic name
+            # or the method specific name.
+            DependencyMethods.EXTRAFRAMEWORK: functools.partial(framework_class, framework_name or name),
+            DependencyMethods.PKGCONFIG: functools.partial(pkgconfig_class, pkgconfig_name or name),
+            DependencyMethods.CMAKE: functools.partial(cmake_class, cmake_name or name),
+            DependencyMethods.SYSTEM: functools.partial(system_class, name),
+            DependencyMethods.CONFIG_TOOL: None,
```
Comment (jon-turney) [inline mesonbuild/dependencies/base.py]: Maybe this could be written as iterating over the non-obsolete dependency methods, rather than hardcoding a list? Unclear if that would be worthwhile, or how often that list might change...

File: mesonbuild/dependencies/__init__.py [hunk 13]
```diff
@@ -18,58 +18,62 @@
...
+    BlocksDependency, OpenMPDependency, cups_factory, curses_factory, gpgme_factory,
+    libgcrypt_factory, libwmf_factory, netcdf_factory, pcap_factory, python3_factory,
+    shaderc_factory, threads_factory,
+)
 from .platform import AppleFrameworks
-from .ui import GLDependency, GnuStepDependency, Qt4Dependency, Qt5Dependency, SDL2Dependency, WxDependency, VulkanDependency
+from .ui import GnuStepDependency, Qt4Dependency, Qt5Dependency, WxDependency, gl_factory, sdl2_factory, vulkan_factory
 
 
 packages.update({
```
Comment (jon-turney) [inline mesonbuild/dependencies/__init__.py]: It might be useful to have a comment here about what the values in this dictionary are supposed to be: subclasses of Dependency, or subclasses of DependencyFactory, or things with a compatible constructor signature to that...

File: mesonbuild/dependencies/base.py (hunk 12 above)
Comment (dcbaker) [reply mesonbuild/dependencies/base.py]: I'm not sure if it can, or if the resulting code would really be any more maintainable, given that the methods are invoked differently.

**Comment by dcbaker** [comment]
@jon-turney, I tried to address all of your comments in this latest version.

File: mesonbuild/dependencies/__init__.py [hunk 14]
```diff
@@ -33,6 +33,10 @@
 from .ui import GnuStepDependency, Qt4Dependency, Qt5Dependency, WxDependency, gl_factory, sdl2_factory, vulkan_factory
 
 
+# This is a dict where the keys should be strings, and tthe values must be one of:
```
Comment (jon-turney) [inline mesonbuild/dependencies/__init__.py]: tthe -> the

**Comment by jon-turney** [comment]
I'll see if I can reproduce the failure macOS, if I have time.

**Comment by dcbaker** [comment]
@jon-turney, thanks! that turned out to be exactly the problem. I was calling `pop` where I really should have been calling `get`

**Comment by dcbaker** [comment]
@scivision (or anyone else) Is this ready to land?

File: mesonbuild/dependencies/base.py [hunk 15]
```diff
@@ -2173,6 +2105,83 @@ def log_tried(self):
...
+
+        By default this only remove EXTRAFRAMEWORK dependencies for non-mac platforms.
+        """
+        # Extra frameworks are only valid for macOS and other apple products
+        if (method is DependencyMethods.EXTRAFRAMEWORK and
+                not env.machines[for_machine].is_darwin()):
+            return False
+        return True
+
+    def __call__(self, env: Environment, for_machine: MachineChoice,
```
Comment (mensinda) [inline mesonbuild/dependencies/base.py]: There should be a unit test checking the order of the returned methods. This can lead to really messy bugs if this breaks for some reason.

File: mesonbuild/dependencies/base.py [hunk 16]
```diff
@@ -2173,6 +2105,83 @@ def log_tried(self):
...
+    :*_class: A *type* or callable that creates a class, and has the
+        signature of an ExternalDependency
+    :system_class: If you pass DependencyMethods.SYSTEM in methods, you must
+        set this argument.
+    """
+
+    def __init__(self, name: str, methods: T.List[DependencyMethods], *,
+                 extra_kwargs: T.Optional[T.Dict[str, T.Any]] = None,
+                 pkgconfig_name: T.Optional[str] = None,
+                 pkgconfig_class: 'T.Type[PkgConfigDependency]' = PkgConfigDependency,
```
[the remark this answers is not in the document]
Comment (dcbaker) [reply mesonbuild/dependencies/base.py]: I really prefer the hardcoded list, it's less typing and it makes the type system cleaner I think.

One of the things I really wanted to do here was make an interface, not a subclassing system. The DependencyFactory is meant to handle the simple, common, cases: one dependency has multiple discovery methods, such as a system method on macOS and PkgConfig on Linux. Then you can write a simple, clean easy to understand function for cases that don't fit that simple case, IMO that makes for simpler, easier to understand code all around.

File: mesonbuild/dependencies/base.py (hunk 15 above)
Comment (dcbaker) [reply mesonbuild/dependencies/base.py]: That's a good call, I've added one locally

**Comment by dcbaker** [comment]
Your example doesn't actually work, because the kwargs at that point aren't the kwargs coming from the user, they're kwargs that the dependency needs to set. The only case of using it (so far) is the python3 ExtraFrameworkDependency, which needs to overwrite the paths to search because the system path will always have python2, and we don't want to find that.

**Comment by mensinda** [comment]
I get your point with the hardcoded kwargs and the overall design decisions. However, I still don't like the code duplication for the processing the dependency methods in the free factory functions. Since subclassing doesn't work and/or is a pain, maybe we could use a decorator instead?

Something like:

```python
@depndency_methods([DependencyMethods.PKGCONFIG, DependencyMethods.CMAKE])
def netcdf_factory(env: 'Environment', for_machine: 'MachineChoice', kwargs: T.Dict[str, T.Any],
                   methods: T.List[DependencyMethods]) -> T.List['DependencyType']:
    candidates = []  # type: T.List['DependencyType']

    if DependencyMethods.PKGCONFIG in methods:
        # Add pkg-config

    if DependencyMethods.CMAKE in methods:
        # Add CMake

    return candidates
```

This would make the intent of the code more readable and it is harder to mess up in the future IMHO.

File: run_unittests.py [hunk 17]
```diff
@@ -1194,6 +1195,17 @@ def test_sort_libpaths(self):
             ['/home/mesonuser/.local/lib/pkgconfig', '/usr/local/libdata/pkgconfig']),
             ['/home/mesonuser/.local/lib', '/usr/local/lib', '/usr/lib'])
 
+    def test_dependency_factory_order(self):
+        b = mesonbuild.dependencies.base
+        f = b.DependencyFactory(
+            'test_dep',
+            methods=[b.DependencyMethods.PKGCONFIG, b.DependencyMethods.CMAKE]
+        )
```
Comment (mensinda) [inline run_unittests.py]: Please also add a second `DependencyFactory` where the methods are reversed.

**Comment by mensinda** [comment]
This looks good to me, besides the minor complaint from above and can be merged once the CI is green.

File: run_unittests.py (hunk 17 above)
Comment (dcbaker) [reply run_unittests.py]: Okay, I think I have the test correct now, and I'm not rebasing out merges.
